### PR TITLE
Add PHP 8.4 support to Smarty

### PR DESCRIPTION
**Pull request by jolelievre**
This PR is mostly a backport of https://github.com/smarty-php/smarty/pull/1043 for version 4

This pull request updates the code base to support the deprecations introduced in PHP 8.4. It should be backwards compatible.

Updated the function signatures that had implicit null assignment to typed parameters.
Added PHP 8.4 to the 'run tests'-shell script.
Bumped the PHP version in the README.md

### CODE REVIEW

File: libs/sysplugins/smartycompilerexception.php [hunk 1]
```diff
@@ -14,14 +14,14 @@ class SmartyCompilerException extends SmartyException
...
-     * @param Throwable|null $previous The previous exception used for the exception chaining.
+     * @param \Throwable|null $previous The previous exception used for the exception chaining.
      */
     public function __construct(
         string $message = "",
         int $code = 0,
         ?string $filename = null,
         ?int $line = null,
-        Throwable $previous = null
+        ?\Throwable $previous = null
```
Comment (wisskid) [inline libs/sysplugins/smartycompilerexception.php:24]: why did you add the backslash?

**Comment by wisskid** [comment]
@jolelievre couple of issues with the unit tests, I'm afraid.

File: libs/sysplugins/smartycompilerexception.php (hunk 1 above)
Comment (jolelievre) [reply libs/sysplugins/smartycompilerexception.php:24]: Mostly because it was also added in the PR for V5 https://github.com/smarty-php/smarty/pull/1043/files#diff-d0f40f666506e3868faaf237c1f5a7731be2cd3c96bfa986b564f7c765b42bb5R26

But I can remove it

**Comment by jolelievre** [comment]
Ok thanks @wisskid I'll look into it

File: libs/sysplugins/smartycompilerexception.php (hunk 1 above)
Comment (jolelievre) [reply libs/sysplugins/smartycompilerexception.php:24]: Ah but I'm dumb there's no namespace in smarty 4, so not needed

**Comment by jolelievre** [comment]
@wisskid it should be good now, at least locally I don't have any errors anymore, I had missed some E_STRICT
And I remove the `\` for the `Throwable`

**Review by wisskid — APPROVED** [review]

**Comment by jolelievre** [comment]
Thanks @wisskid
